File: app_canibuythat/src/main/java/com/gb/canibuythat/exception/DomainException.java
package com.gb.canibuythat.exception;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.HttpException;

public class DomainException extends Throwable {

    private ResponseBody responseBody;
    private int code;
    private Kind kind;
    private Action action;

    public DomainException(Throwable raw) {
        this(raw.getMessage(), raw);
    }

    public DomainException(String message) {
        this(message, null);
    }

    public DomainException(String message, Throwable cause) {
        super(message, cause);
        this.code = -1;
        this.kind = Kind.GENERIC;

        if (cause instanceof HttpException) {
            this.kind = Kind.HTTP;
            this.code = ((HttpException) cause).code();

            if (code == 401) {
                action = Action.LOGIN;
            }

            this.responseBody = ((HttpException) cause).response().errorBody();
        }
        if (cause instanceof IOException) {
            this.kind = Kind.NETWORK;
        }
    }

    public ResponseBody getResponseBody() {
        return responseBody;
    }

    public int getCode() {
        return code;
    }

    public Kind getKind() {
        return kind;
    }

    public Action getAction() {
        return action;
    }

    public enum Kind {
        NETWORK,
        HTTP,
        GENERIC;
    }

    public enum Action {
        LOGIN
    }
}

File: core/src/main/java/com/gb/canibuyit/error/DomainException.kt
package com.gb.canibuyit.error

import okhttp3.ResponseBody
import retrofit2.HttpException
import java.io.IOException

open class DomainException @JvmOverloads constructor(
    message: String?,
    cause: Throwable? = null) : Throwable(message, cause) {

    var responseBody: ResponseBody? = null
        private set

    var code: Int
        private set

    var kind: Kind
        private set

    var action: Action? = null
        private set

    constructor(raw: Throwable) : this(raw.message, raw)

    init {
        this.code = -1
        this.kind = Kind.GENERIC

        if (cause is HttpException) {
            this.kind = Kind.HTTP
            this.code = cause.code()

            if (code == 400) {
                action = Action.LOGIN
            }
            this.responseBody = cause.response()?.errorBody()
        }
        if (cause is IOException) {
            this.kind = Kind.NETWORK
        }
    }

    enum class Kind {
        NETWORK,
        HTTP,
        GENERIC
    }

    enum class Action {
        LOGIN
    }
}
